MRY record multilevel caching

mlaflamm opened the pull request.

appaquet committed:
Exprimental first level caching

Commit:
Merge branch 'master' into manu-multilevel-caching

Commit:
First draft of multilevel caching

Commit:
Working multilevel caching

Commit:
Repackaging

Commit:
Split tests

Commit:
MysqlStorage cache IT tests

alexbergeron commented on mry-core/src/it/scala/com/wajam/mry/storage/mysql/MysqlStorageFixture.scala:
```diff
@@ -0,0 +1,72 @@
+import com.wajam.mry.storage.Storage
+import com.wajam.nrv.service.TokenRange
+import com.wajam.nrv.utils.timestamp.Timestamp
+import com.wajam.mry.execution.{ExecutionContext, Value, Transaction}
```
We should plan to integrate scalariform in MRY's build

mlaflamm replied:
`eventually { scalariformEnabled == true }`

oliviertoupin replied:
JIRA ticket?

alexbergeron commented on mry-core/src/it/scala/com/wajam/mry/storage/mysql/MysqlStorageFixture.scala:
```diff
@@ -0,0 +1,72 @@
+      mysqlStorage.syncModel(model)
+      mysqlStorage.start()
+      mysqlStorage.setCurrentConsistentTimestamp((_) => currentConsistentTimestamp)
+      val theFixture = FixtureParam(mysqlStorage, Map("mysql" -> mysqlStorage))
```
Why `theFixture`? I don't see any name collision, and even then I don't think it would be a good name.

mlaflamm replied:
I agree and I will rename. The name originate from the [fixture.FunSuite](http://doc.scalatest.org/2.1.0/index.html#org.scalatest.fixture.FunSuite) examples.

alexbergeron commented on mry-core/src/it/scala/com/wajam/mry/storage/mysql/MysqlStorageFixture.scala:
```diff
@@ -0,0 +1,72 @@
+  def withFixture(test: (FixtureParam) => Unit)(implicit createStorage: () => MysqlStorage = createMysqlStorage _) = {
+
+    // create the fixture
+    var mysqlStorage: MysqlStorage = createStorage()
```
Could be a val

mlaflamm commented:
@appaquet @alexbergeron: Should I put the caching configuration into [MysqlStorageConfiguration](https://github.com/wajam/mry/blob/master/mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala#L872-L884) even if the caching is a mix-in trait?

alexbergeron commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/Model.scala:
```diff
@@ -49,6 +51,14 @@ class Table(val name: String, parent: Option[Table] = None, val maxVersions: Int
+      case Some(tableParent) => getTopLevelTable(tableParent)
+      case None => table
+    }
+  }
```
I would love to try the following implementation, from what I've herd, it might also be tail recursive

``` scala
@tailrec
final def getTopLevelTable(): Table = {
  this.parentTable match {
    case Some(tableParent) => tableParent.getTopLevelTable()
    case None => this
  }
}
```

alexbergeron commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala:
```diff
@@ -866,12 +866,12 @@ object MysqlTransaction {
 }
 
-case class AccessKey(var key: String)
+case class AccessKey(key: String)
```
I love that change!

alexbergeron commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/cache/TransactionCache.scala:
```diff
@@ -0,0 +1,151 @@
+        trxTableCaches += topLevelTable -> cache
+        cache
+      }
+    }
```
``` scala
trxTableCaches.get(topLevelTable) getOrElse {
  val cache = new TransactionTableCache()
  trxTableCaches += topLevelTable -> cache
  cache
}
```

Commit:
Proper cache config

alexbergeron commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/cache/TransactionCache.scala:
```diff
@@ -0,0 +1,151 @@
+
+    def toIterable: Iterable[(AccessPath, CachedValue)] = {
+      import collection.JavaConversions._
+      cache.entrySet().toIterator.toIterable.map(e => e.getKey -> e.getValue)
```
I love that `.toIterator.toIterable` part!

(Ok, I don't but I don't have any better suggestions that come to my mind now)

mlaflamm replied:
This ensure true laziness.

alexbergeron replied:
Define true laziness?

Looking at it, a view of the collection might suffice.

mlaflamm replied:
Now I remember. The original `Set` is a java `TreeSet`, thus ordered, but `toIterable` was making a copy in a normal scala `Set` thus not ordered any more. So there was two issues: not lazy and not ordered.

alexbergeron replied:
Do you want to preserve order, then?

mlaflamm replied:
Yes this is critical. We must transfer the ancestor records to the storage cache before the descendants.

alexbergeron replied:
I see...

Still, are you sure that calling `toIterable` transforms it into a type that presevres laziness?

mlaflamm replied:
Thinking more about it this might not be important any more! The initial design used to always invalidate descendants on update but now descendants are only invalidated on ancestor invalidation.

mlaflamm replied:
About laziness, `toIterable` returns a `Stream` in this situation.

alexbergeron replied:
Good then!

alexbergeron commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/cache/TransactionCache.scala:
```diff
@@ -0,0 +1,151 @@
+        getIfPresent(parentPath) match {
+          case Some(CachedValue(None, Action.Put)) => true
+          case _ => isAncestorDeleted(parentPath)
+        }
```
``` scala
(getIfPresent(parentPath) == Some(CachedValue(None, Action.Put))) || isAncestorDeleted(parentPath)
```

alexbergeron commented on mry-core/src/test/scala/com/wajam/mry/storage/mysql/cache/CacheSetup.scala:
```diff
@@ -0,0 +1,66 @@
+package com.wajam.mry.storage.mysql.cache
+
+import language.implicitConversions
```
I like putting `scala.` in front of my language imports.

(Also, I realized at Nescala that many other devs/companies just add theses as compiler flags, using something akin to

``` scala
scalacOptions ++= Seq( "-language:implicitConversions", "-language:postfixOps", "-language:higherKinds", "-language:existentials" )
```

in their `Project.scala`. I'm fine with most, if not all (We've never actually had any problems with `postfixOps`))

alexbergeron replied:
(higherKinds and existentials being here for my happiness, through I'm surprised I'm the only one who had to use existentials at this point)

alexbergeron commented on mry-core/src/test/scala/com/wajam/mry/storage/mysql/cache/CacheSetup.scala:
```diff
@@ -0,0 +1,66 @@
+import com.wajam.mry.storage.mysql._
+import org.scalatest.Matchers._
+import com.wajam.mry.storage.mysql.AccessPath
+import com.wajam.mry.storage.mysql.AccessKey
```
Useless imports

alexbergeron replied:
(Already imported by `import com.wajam.mry.storage.mysql._`)

alexbergeron commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/cache/HierarchicalCache.scala:
```diff
@@ -0,0 +1,74 @@
+import java.util.concurrent.{ConcurrentSkipListSet, TimeUnit}
+import com.wajam.commons.Logging
+import com.wajam.mry.storage.mysql._
+import com.wajam.mry.storage.mysql.AccessPath
```
Useless import

Commit:
Multiple small improvements raised from code review

jgagnon1 commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/cache/HierarchicalCache.scala:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private object CacheRemovalListener extends RemovalListener[AccessPath, Record] {
+    def onRemoval(notification: RemovalNotification[AccessPath, Record]) {
```
this notation is deprecated it Scala 2.11 use `: Unit =`

alexbergeron replied:
Not sure if it's already deprecated, but it's definitively in a long-term plan to deprecate then remove support of this notation

jgagnon1 commented on mry-core/src/main/scala/com/wajam/mry/storage/mysql/cache/HierarchicalCache.scala:
```diff
@@ -0,0 +1,73 @@
+      }
+    }
+
+    private def wasEvicted(cause: RemovalCause) = cause != RemovalCause.EXPLICIT && cause != RemovalCause.REPLACED
```
deprecated notation

alexbergeron replied:
Where?

jgagnon1 replied:
I swear it was there 2minutes ago :) .. no my bad, please ignore.

Commit:
Disable cache for replicated data read/written on slave replicas

Commit:
Add cache metrics + record ressurection bug fix

Commit:
Change after review

mlaflamm committed:
Merge pull request #160 from wajam/manu-caching-metrics

Add cache metrics

Commit:
Metrics now per top ancestor OR per last decendants (could be both before)

Commit:
Merge branch 'manu-multilevel-caching' into manu-replication-caching

Commit:
Atomic hierarchical cache invalidateAll() implementation

mlaflamm committed:
Merge pull request #161 from wajam/manu-replication-caching

Disable cache for slave member

Commit:
Cleanup leftovers

alexbergeron commented:
+1

Commit:
Do not flush transaction cache if commited after the cache is invalidated

When a service member goes Down while a transaction is running, the transaction cache is not flush in the current storage cache if committed after the storage cache is invalidated.

alexbergeron commented:
+1

oliviertoupin commented on mry-core/src/it/scala/com/wajam/mry/storage/mysql/MysqlStorageFixture.scala:
```diff
@@ -0,0 +1,65 @@
+  val table2_1 = table2.addTable(new Table("table2_1"))
+  val table2_1_1 = table2_1.addTable(new Table("table2_1_1"))
+
+  val defaultConfig = MysqlStorageConfiguration("mysql", "localhost", "mry", "mry", "mry",
```
Don't we have a test config somewhere already with those name?

mlaflamm replied:
Yes in IT tests but unfortunately fixtures are not sharable between unit an
IT tests...

appaquet commented:
+1
